numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -0,0 +1,3035 @@
+                   &string_bool_output_resolve_descriptors,
+                   &string_isnan_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isalpha", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isalpha_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isdecimal", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isdecimal_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isdigit", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isdigit_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isnumeric", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isnumeric_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isspace", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isspace_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    PyArray_DTypeMeta *intp_output_dtypes[] = {
+        &PyArray_StringDType,
+        &PyArray_IntpDType
+    };
+
+    if (init_ufunc(umath, "str_len", intp_output_dtypes,
+                   &string_intp_output_resolve_descriptors,
+                   &string_strlen_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    PyArray_DTypeMeta *binary_dtypes[] = {
+            &PyArray_StringDType,
+            &PyArray_StringDType,
+            &PyArray_StringDType,
+    };
+
+    if (init_ufunc(umath, "maximum", binary_dtypes, binary_resolve_descriptors,
+                   &maximum_strided_loop, 2, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "minimum", binary_dtypes, binary_resolve_descriptors,
+                   &minimum_strided_loop, 2, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "add", binary_dtypes, binary_resolve_descriptors,
+                   &add_strided_loop, 2, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    INIT_MULTIPLY(Int64, int64);
+    INIT_MULTIPLY(UInt64, uint64);
+// TODO: for reasons I don't understand at the moment we can't add these using
```

Comment:
@seberg - any advice on why the promotion doesn't "just work" without explicitly adding promotors for every integer type?

Reply:
I am not sure what @ngoldbaum tried.  I assume it refers to the code below, which is indeed quite annoying!

I would think what works is adding a promoter for `&PyArray_StringDType, &PyArray_DescrType` (or `None` I think might work).  Which will then be called always if one of the two dtypes is string dtype (and it must finish the promotion).

Reply:
Can one put in the more general abstract integer and unsigned integer types?

Reply:
the abstract integer dtype is only for python integers. We don't have a way of representing the dtype type hierarchy in the C DType API at the moment.

The promoter sebastian is suggesting will only fire for `multiply`, so the fact that it will fire for all dtypes is fine because multiplying by all other types is an error. Of course that limitation would be more annoying for other dtypes.

Reply:
Yes, I always meant to create an abstract base DType you can easily register with that matches any integer/float, etc.
For some reason it was a bit trickier than I thought it would be (I don't recall now).  The machinery should support that just fine (I am not 100% sure we might have ripped the logic out because it was unused at the time).

At some point I thought the Python abstract integer DType could serve that purpose also, but I don't think it can anymore (we need to match exactly for some of the more complex stuff).

Reply:
OK, thanks! And, yes, the promoter with `None` sounds like a good idea as well!